sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Azure.Communication.CallAutomation
+{
+    /// <summary>
+    /// Event backlog that saves event for cases where events arrive earlier than the response of the method call.
+    /// </summary>
+    internal class EventBacklog
+    {
+        private const int DEFAULT_TIMEOUT = 5;
+        private const int MAXIMUM_EVENTBACKLOGS_AT_ONCE = 10000;
+
+        private TimeSpan _expiringTimeout;
+
+        private ConcurrentDictionary<string, (CallAutomationEventBase, Timer)> _eventBacklog;
+
+        internal EventBacklog(TimeSpan defaultExpiringTimeout = default)
+        {
+            _expiringTimeout = defaultExpiringTimeout == default ? TimeSpan.FromSeconds(DEFAULT_TIMEOUT) : defaultExpiringTimeout;
+            _eventBacklog = new ConcurrentDictionary<string, (CallAutomationEventBase, Timer)>();
+        }
+
+        /// <summary>
+        /// Adds event to EventsBacklog to be retrieved.
+        /// </summary>
+        /// <param name="backlogEventId">Internally used id for tracking the saved event.</param>
+        /// <param name="eventsToBeSaved">Incoming Event to be saved.</param>
+        /// <returns></returns>
+        internal bool AddEvent(string backlogEventId, CallAutomationEventBase eventsToBeSaved)
+        {
+            if (_eventBacklog.Count < MAXIMUM_EVENTBACKLOGS_AT_ONCE)
+            {
+                // Set Timer for this events. Expires after awhile and deletes itself from the Dictionary.
+                var expiringTimer = new Timer(new TimerCallback(TimerProc), backlogEventId, (int)_expiringTimeout.TotalMilliseconds, 0);
+
+                return _eventBacklog.TryAdd(backlogEventId, (eventsToBeSaved, expiringTimer));
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        internal bool GetAndRemoveEvent(IEnumerable<Type> eventTypes, string callConnectionId, string operationContext, out KeyValuePair<string, CallAutomationEventBase> matchingEvent)
```

Comment:
This is a typical example for Try Pattern. The method should be prefixed with `Try`